<a id="load"></a>
## 1. Load and explore claims data

The csv file `resubmissionClaims.csv` Load the file to Apache® Spark DataFrame using code below.

In [3]:
# The code was removed by Watson Studio for sharing.

[Row(DESYNPUF_ID='00016F745862898F', CLM_ID='5.42E+14', SEGMENT='1', CLM_FROM_DT='20090602', CLM_THRU_DT='20090602', PRVDR_NUM='3901GS', CLM_PMT_AMT='30', NCH_PRMRY_PYR_CLM_PD_AMT='0', AT_PHYSN_NPI='2963419753', OP_PHYSN_NPI=None, OT_PHYSN_NPI='2963419753', NCH_BENE_BLOOD_DDCTBL_LBLTY_AM='0', ICD9_DGNS_CD_1='V5832', ICD9_DGNS_CD_2='V5861', ICD9_DGNS_CD_3='2724', ICD9_DGNS_CD_4='3182', ICD9_DGNS_CD_5='V5869', ICD9_DGNS_CD_6='42731', ICD9_DGNS_CD_7=None, ICD9_DGNS_CD_8=None, ICD9_DGNS_CD_9=None, ICD9_DGNS_CD_10=None, ICD9_PRCDR_CD_1=None, ICD9_PRCDR_CD_2=None, ICD9_PRCDR_CD_3=None, ICD9_PRCDR_CD_4=None, ICD9_PRCDR_CD_5=None, ICD9_PRCDR_CD_6=None, NCH_BENE_PTB_DDCTBL_AMT='0', NCH_BENE_PTB_COINSRNC_AMT='0', ADMTNG_ICD9_DGNS_CD=None, HCPCS_CD_1='85610', HCPCS_CD_2='80048', HCPCS_CD_3='80061', HCPCS_CD_4='82306', HCPCS_CD_5='96372', HCPCS_CD_6='87088', HCPCS_CD_7='85025', HCPCS_CD_8='80076', HCPCS_CD_9='84075', HCPCS_CD_10=None, HCPCS_CD_11=None, HCPCS_CD_12=None, HCPCS_CD_13=None, HCPCS_CD_

In [4]:
df_data.printSchema()

root
 |-- DESYNPUF_ID: string (nullable = true)
 |-- CLM_ID: string (nullable = true)
 |-- SEGMENT: string (nullable = true)
 |-- CLM_FROM_DT: string (nullable = true)
 |-- CLM_THRU_DT: string (nullable = true)
 |-- PRVDR_NUM: string (nullable = true)
 |-- CLM_PMT_AMT: string (nullable = true)
 |-- NCH_PRMRY_PYR_CLM_PD_AMT: string (nullable = true)
 |-- AT_PHYSN_NPI: string (nullable = true)
 |-- OP_PHYSN_NPI: string (nullable = true)
 |-- OT_PHYSN_NPI: string (nullable = true)
 |-- NCH_BENE_BLOOD_DDCTBL_LBLTY_AM: string (nullable = true)
 |-- ICD9_DGNS_CD_1: string (nullable = true)
 |-- ICD9_DGNS_CD_2: string (nullable = true)
 |-- ICD9_DGNS_CD_3: string (nullable = true)
 |-- ICD9_DGNS_CD_4: string (nullable = true)
 |-- ICD9_DGNS_CD_5: string (nullable = true)
 |-- ICD9_DGNS_CD_6: string (nullable = true)
 |-- ICD9_DGNS_CD_7: string (nullable = true)
 |-- ICD9_DGNS_CD_8: string (nullable = true)
 |-- ICD9_DGNS_CD_9: string (nullable = true)
 |-- ICD9_DGNS_CD_10: string (nullable = 

In [5]:
df_data.take(1)

[Row(DESYNPUF_ID='00016F745862898F', CLM_ID='5.42E+14', SEGMENT='1', CLM_FROM_DT='20090602', CLM_THRU_DT='20090602', PRVDR_NUM='3901GS', CLM_PMT_AMT='30', NCH_PRMRY_PYR_CLM_PD_AMT='0', AT_PHYSN_NPI='2963419753', OP_PHYSN_NPI=None, OT_PHYSN_NPI='2963419753', NCH_BENE_BLOOD_DDCTBL_LBLTY_AM='0', ICD9_DGNS_CD_1='V5832', ICD9_DGNS_CD_2='V5861', ICD9_DGNS_CD_3='2724', ICD9_DGNS_CD_4='3182', ICD9_DGNS_CD_5='V5869', ICD9_DGNS_CD_6='42731', ICD9_DGNS_CD_7=None, ICD9_DGNS_CD_8=None, ICD9_DGNS_CD_9=None, ICD9_DGNS_CD_10=None, ICD9_PRCDR_CD_1=None, ICD9_PRCDR_CD_2=None, ICD9_PRCDR_CD_3=None, ICD9_PRCDR_CD_4=None, ICD9_PRCDR_CD_5=None, ICD9_PRCDR_CD_6=None, NCH_BENE_PTB_DDCTBL_AMT='0', NCH_BENE_PTB_COINSRNC_AMT='0', ADMTNG_ICD9_DGNS_CD=None, HCPCS_CD_1='85610', HCPCS_CD_2='80048', HCPCS_CD_3='80061', HCPCS_CD_4='82306', HCPCS_CD_5='96372', HCPCS_CD_6='87088', HCPCS_CD_7='85025', HCPCS_CD_8='80076', HCPCS_CD_9='84075', HCPCS_CD_10=None, HCPCS_CD_11=None, HCPCS_CD_12=None, HCPCS_CD_13=None, HCPCS_CD_

In [7]:
df_selected = df_data.select("SEGMENT","PRVDR_NUM","CLM_PMT_AMT","HCPCS_CD_1","HCPCS_CD_2","HCPCS_CD_3","HCPCS_CD_4", 
                             "HCPCS_CD_5", "HCPCS_CD_6", "HCPCS_CD_7", "HCPCS_CD_8" , "HCPCS_CD_9", "HCPCS_CD_10",
                             "HCPCS_CD_11",
                             "RESUBMIT")

from pyspark.sql.functions import col, when
k = col("SEGMENT")
df_selected = df_selected.withColumn("SEGMENT", when(k.isNull(), "0").otherwise(k))

from pyspark.sql.functions import col, when
k = col("PRVDR_NUM")
df_selected = df_selected.withColumn("PRVDR_NUM", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("CLM_PMT_AMT")
df_selected = df_selected.withColumn("CLM_PMT_AMT", when(k.isNull(), "0").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_1")
df_selected = df_selected.withColumn("HCPCS_CD_1", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_2")
df_selected = df_selected.withColumn("HCPCS_CD_2", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_3")
df_selected = df_selected.withColumn("HCPCS_CD_3", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_4")
df_selected = df_selected.withColumn("HCPCS_CD_4", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_5")
df_selected = df_selected.withColumn("HCPCS_CD_5", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_6")
df_selected = df_selected.withColumn("HCPCS_CD_6", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_7")
df_selected = df_selected.withColumn("HCPCS_CD_7", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_8")
df_selected = df_selected.withColumn("HCPCS_CD_8", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_9")
df_selected = df_selected.withColumn("HCPCS_CD_9", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_10")
df_selected = df_selected.withColumn("HCPCS_CD_10", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("HCPCS_CD_11")
df_selected = df_selected.withColumn("HCPCS_CD_11", when(k.isNull(), "__NA__").otherwise(k))

df_selected.cache()
df_selected.persist()
df_selected.take(5)

[Row(SEGMENT='1', PRVDR_NUM='3901GS', CLM_PMT_AMT='30', HCPCS_CD_1='85610', HCPCS_CD_2='80048', HCPCS_CD_3='80061', HCPCS_CD_4='82306', HCPCS_CD_5='96372', HCPCS_CD_6='87088', HCPCS_CD_7='85025', HCPCS_CD_8='80076', HCPCS_CD_9='84075', HCPCS_CD_10='__NA__', HCPCS_CD_11='__NA__', RESUBMIT='N'),
 Row(SEGMENT='1', PRVDR_NUM='3939PG', CLM_PMT_AMT='30', HCPCS_CD_1='71101', HCPCS_CD_2='78480', HCPCS_CD_3='94060', HCPCS_CD_4='__NA__', HCPCS_CD_5='__NA__', HCPCS_CD_6='__NA__', HCPCS_CD_7='__NA__', HCPCS_CD_8='__NA__', HCPCS_CD_9='__NA__', HCPCS_CD_10='__NA__', HCPCS_CD_11='__NA__', RESUBMIT='N'),
 Row(SEGMENT='2', PRVDR_NUM='4600BC', CLM_PMT_AMT='1700', HCPCS_CD_1='A4657', HCPCS_CD_2='__NA__', HCPCS_CD_3='__NA__', HCPCS_CD_4='__NA__', HCPCS_CD_5='__NA__', HCPCS_CD_6='__NA__', HCPCS_CD_7='__NA__', HCPCS_CD_8='__NA__', HCPCS_CD_9='__NA__', HCPCS_CD_10='__NA__', HCPCS_CD_11='__NA__', RESUBMIT='N'),
 Row(SEGMENT='2', PRVDR_NUM='4600BC', CLM_PMT_AMT='3200', HCPCS_CD_1='A4657', HCPCS_CD_2='__NA__', 

Explore the loaded data by using the following Apache® Spark DataFrame methods:
-  print schema
-  print top ten records
-  count all records

As you can see, the data contains various fields. Resubmssion field is the one we would like to predict (label).

In [8]:
df_selected.printSchema()

root
 |-- SEGMENT: string (nullable = true)
 |-- PRVDR_NUM: string (nullable = true)
 |-- CLM_PMT_AMT: string (nullable = true)
 |-- HCPCS_CD_1: string (nullable = true)
 |-- HCPCS_CD_2: string (nullable = true)
 |-- HCPCS_CD_3: string (nullable = true)
 |-- HCPCS_CD_4: string (nullable = true)
 |-- HCPCS_CD_5: string (nullable = true)
 |-- HCPCS_CD_6: string (nullable = true)
 |-- HCPCS_CD_7: string (nullable = true)
 |-- HCPCS_CD_8: string (nullable = true)
 |-- HCPCS_CD_9: string (nullable = true)
 |-- HCPCS_CD_10: string (nullable = true)
 |-- HCPCS_CD_11: string (nullable = true)
 |-- RESUBMIT: string (nullable = true)



In [9]:
df_selected.count()

67834

As you can see, the data set contains 67834 records.

<a id="model"></a>
## 3. Create an Apache® Spark machine learning model

In this section you will learn how to prepare data, create an Apache® Spark machine learning pipeline, and train a model.

### 3.1: Prepare data

Split your data into the train, test, and predict data sets:

In [10]:
splitted_data = df_selected.randomSplit([80.0,18.0,2.0])
train_data = splitted_data[0]
test_data = splitted_data[1]
predict_data = splitted_data[2]

test_data.cache()
train_data.cache()
predict_data.cache()

print ("Number of training records: " , str(train_data.count()))
print ("Number of testing records : " , str(test_data.count()))
print ("Number of predict records : " + str(predict_data.count()))



Number of training records:  54111
Number of testing records :  12387
Number of predict records : 1336


The data has been split into three data sets: 

-  The train data set, which is the largest group, is used for training.
-  The test data set is used for model evaluation and to test the assumptions of the model.
-  The predict data set is used for predictions.

### 3.2: Create pipeline and train a model

In this section you will create an Apache® Spark machine learning pipeline and then train the model.

Import the Apache® Spark machine learning packages that you'll need in the subsequent steps:

In [11]:
from pyspark.ml.feature import OneHotEncoder, StringIndexer, IndexToString, VectorAssembler
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.evaluation import BinaryClassificationEvaluator
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml import Pipeline, Model

Convert all the string fields to numeric ones by using the StringIndexer transformer:

In [13]:
stringIndexer_segment = StringIndexer(inputCol="SEGMENT", outputCol="SEGMENT_IX").setHandleInvalid("skip")
stringIndexer_prvdr_num = StringIndexer(inputCol="PRVDR_NUM", outputCol="PRVDR_NUM_IX").setHandleInvalid("skip")
stringIndexer_clm_pmt_amt = StringIndexer(inputCol="CLM_PMT_AMT", outputCol="CLM_PMT_AMT_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_1 = StringIndexer(inputCol="HCPCS_CD_1", outputCol="HCPCS_CD_1_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_2 = StringIndexer(inputCol="HCPCS_CD_2", outputCol="HCPCS_CD_2_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_3 = StringIndexer(inputCol="HCPCS_CD_3", outputCol="HCPCS_CD_3_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_4 = StringIndexer(inputCol="HCPCS_CD_4", outputCol="HCPCS_CD_4_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_5 = StringIndexer(inputCol="HCPCS_CD_5", outputCol="HCPCS_CD_5_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_6 = StringIndexer(inputCol="HCPCS_CD_6", outputCol="HCPCS_CD_6_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_7 = StringIndexer(inputCol="HCPCS_CD_7", outputCol="HCPCS_CD_7_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_8 = StringIndexer(inputCol="HCPCS_CD_8", outputCol="HCPCS_CD_8_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_9 = StringIndexer(inputCol="HCPCS_CD_9", outputCol="HCPCS_CD_9_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_10 = StringIndexer(inputCol="HCPCS_CD_10", outputCol="HCPCS_CD_10_IX").setHandleInvalid("skip")
stringIndexer_hpcscd_11 = StringIndexer(inputCol="HCPCS_CD_11", outputCol="HCPCS_CD_11_IX").setHandleInvalid("skip")
stringIndexer_resubmit = StringIndexer(inputCol="RESUBMIT", outputCol="label").setHandleInvalid("skip")

Create a feature vector by combining all features together:

In [14]:
vectorAssembler = VectorAssembler(inputCols=["SEGMENT_IX", "PRVDR_NUM_IX", "CLM_PMT_AMT_IX", 
                                             "HCPCS_CD_1_IX",
                                            "HCPCS_CD_2_IX",
                                             "HCPCS_CD_3_IX",
                                             "HCPCS_CD_4_IX",
                                             "HCPCS_CD_5_IX",
                                             "HCPCS_CD_6_IX",
                                             "HCPCS_CD_7_IX",
                                             "HCPCS_CD_8_IX",
                                             "HCPCS_CD_9_IX",
                                             "HCPCS_CD_10_IX",
                                             "HCPCS_CD_11_IX",
                                            ], outputCol="features")

Next, define the estimators you want to use for classification. Random Forest is used in the following example.

In [15]:
rf = RandomForestClassifier(labelCol="label", featuresCol="features", maxBins=8192, numTrees=10)

Now build the pipeline. A pipeline consists of transformers and an estimator.

In [16]:
pipeline_rf = Pipeline(stages=[stringIndexer_segment, 
                               stringIndexer_prvdr_num, 
                               stringIndexer_clm_pmt_amt, 
                               stringIndexer_hpcscd_1, 
                               stringIndexer_hpcscd_2, 
                               stringIndexer_hpcscd_3, 
                               stringIndexer_hpcscd_4, 
                               stringIndexer_hpcscd_5, 
                               stringIndexer_hpcscd_6, 
                               stringIndexer_hpcscd_7, 
                               stringIndexer_hpcscd_8, 
                               stringIndexer_hpcscd_9, 
                               stringIndexer_hpcscd_10, 
                               stringIndexer_hpcscd_11, 
                               stringIndexer_resubmit, 
                               vectorAssembler, rf])

Train your Random Forest model by using your pipeline with the train_data set.

In [17]:
model_rf = pipeline_rf.fit(train_data)

You can check your **model accuracy** now. Evaluate the model with the test_data set.

In [18]:
results = model_rf.transform(test_data)

Model Evaluation

In [19]:
test_data.printSchema()

root
 |-- SEGMENT: string (nullable = true)
 |-- PRVDR_NUM: string (nullable = true)
 |-- CLM_PMT_AMT: string (nullable = true)
 |-- HCPCS_CD_1: string (nullable = true)
 |-- HCPCS_CD_2: string (nullable = true)
 |-- HCPCS_CD_3: string (nullable = true)
 |-- HCPCS_CD_4: string (nullable = true)
 |-- HCPCS_CD_5: string (nullable = true)
 |-- HCPCS_CD_6: string (nullable = true)
 |-- HCPCS_CD_7: string (nullable = true)
 |-- HCPCS_CD_8: string (nullable = true)
 |-- HCPCS_CD_9: string (nullable = true)
 |-- HCPCS_CD_10: string (nullable = true)
 |-- HCPCS_CD_11: string (nullable = true)
 |-- RESUBMIT: string (nullable = true)



<a id="persistence"></a>
## 4. Persist the model

In [20]:
selected_results = results.persist()

In [21]:
selected_results.show(10)

+-------+---------+-----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+-----------+-----------+--------+----------+------------+--------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+--------------+--------------+-----+--------------------+--------------------+--------------------+----------+
|SEGMENT|PRVDR_NUM|CLM_PMT_AMT|HCPCS_CD_1|HCPCS_CD_2|HCPCS_CD_3|HCPCS_CD_4|HCPCS_CD_5|HCPCS_CD_6|HCPCS_CD_7|HCPCS_CD_8|HCPCS_CD_9|HCPCS_CD_10|HCPCS_CD_11|RESUBMIT|SEGMENT_IX|PRVDR_NUM_IX|CLM_PMT_AMT_IX|HCPCS_CD_1_IX|HCPCS_CD_2_IX|HCPCS_CD_3_IX|HCPCS_CD_4_IX|HCPCS_CD_5_IX|HCPCS_CD_6_IX|HCPCS_CD_7_IX|HCPCS_CD_8_IX|HCPCS_CD_9_IX|HCPCS_CD_10_IX|HCPCS_CD_11_IX|label|            features|       rawPrediction|         probability|prediction|
+-------+---------+-----------+----------+----------+----------+----------+----------+----------+----------+----------+---

In [29]:
print ("Claim resubmission model precision = {:.9f}%"
       .format(100 - selected_results.filter(selected_results.RESUBMIT == 'Y')
       .filter(selected_results.prediction == 1.0 )
       .count()/float(selected_results.count())) )

Claim resubmission model precision = 99.843208995%


In [30]:
evaluator = BinaryClassificationEvaluator(rawPredictionCol="prediction",labelCol="label",metricName="areaUnderROC")
auc = evaluator.evaluate(selected_results)
print ('Area under curve = {:.2f}.'.format(auc))
print ("Area under curve = ", str(auc*100) + '%')

# Evaluate the model by computing the accuracy & error
mc_evaluator = MulticlassClassificationEvaluator(labelCol="label",predictionCol="prediction", metricName="accuracy")
accuracy = mc_evaluator.evaluate(selected_results)*100
error = 100 - accuracy
print("Accuracy =  " +  str(accuracy) + '%')
print("Test Error = " +  str(error) + '%')

# Evaluate the model by computing the RMSE on the selected data
rg_evaluator = RegressionEvaluator(metricName="rmse", labelCol="label", predictionCol="prediction")
rmse = rg_evaluator.evaluate(selected_results)*100
print("Root-mean-square error = " + str(rmse) + '%')


Area under curve = 1.00.
Area under curve =  99.57667478040004%
Accuracy =  99.28609673389256%
Test Error = 0.7139032661074367%
Root-mean-square error = 8.449279650404776%


In this section you will learn how to store your pipeline and model in Watson Machine Learning repository by using Python client libraries.

First, you must import client libraries.

In [31]:
from repository.mlrepositoryclient import MLRepositoryClient
from repository.mlrepositoryartifact import MLRepositoryArtifact

Put your authentication information from your instance of the Watson Machine Learning service on Bluemix in the next cell. You can find your information on the **Service Credentials** tab of your service instance in Bluemix.

In [32]:
service_path = 'https://us-south.ml.cloud.ibm.com'
username = 'e7ad3869-9346-46ba-a6ab-293b55b99028'
password = '232065f5-2a18-4c91-aa8e-2d5b0d38df7d'
instance_id = '118386ba-432d-4147-94ac-d85978f44550'


Authorize the repository client:

In [33]:
ml_repository_client = MLRepositoryClient(service_path)
ml_repository_client.authorize(username, password)

Create the model artifact (abstraction layer).

In [34]:
model_artifact = MLRepositoryArtifact(model_rf, training_data=train_data, name="Claim Re-submission Prediction Model")

**Tip**: The MLRepositoryArtifact method expects a trained model object, training data, and a model name. (It is this model name that is displayed by the Watson Machine Learning service).

### 4.1: Save the pipeline and model

Save pipeline and model artifacts to your Watson Machine Learning instance:

In [35]:
saved_model = ml_repository_client.models.save(model_artifact)

Get the saved model metadata from Watson Machine Learning. Use the *meta.available_props()* method to get the list of available props.

In [36]:
saved_model.meta.available_props()

dict_keys(['inputDataSchema', 'modelVersionHref', 'pipelineVersionHref', 'evaluationMetrics', 'label', 'runtime', 'creationTime', 'trainingDataSchema', 'authorEmail', 'trainingDataRef', 'evaluationMethod', 'authorName', 'lastUpdated', 'modelType', 'version'])

In [37]:
print ("modelType: " + saved_model.meta.prop("modelType"))
print ("trainingDataSchema: " + str(saved_model.meta.prop("trainingDataSchema")))
print ("creationTime: " + str(saved_model.meta.prop("creationTime")))
print ("modelVersionHref: " + saved_model.meta.prop("modelVersionHref"))
print ("label: " + saved_model.meta.prop("label"))

modelType: sparkml-model-2.3
trainingDataSchema: {'fields': [{'metadata': {}, 'nullable': True, 'name': 'SEGMENT', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'PRVDR_NUM', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'CLM_PMT_AMT', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_1', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_2', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_3', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_4', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_5', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_6', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_7', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_8', 'type': 'string'}, {'metadata': {}, 'nullable': True, 'name': 'HCPCS_CD_9', 'type': 'string'}, {'metadata': {}, 'nullable': T

**Tip**: **modelVersionHref** is our model unique indentifier in the Watson Machine Learning repository.

### 4.2: Load the model

Load the saved model from the specified instance of Watson Machine Learning:

In [38]:
loadedModelArtifact = ml_repository_client.models.get(saved_model.uid)

You can print for example model name to make sure that model has been loaded correctly.

In [39]:
print (str(loadedModelArtifact.name))

Claim Re-submission Prediction Model


As you can see the name is correct. You have learned how save and load the model from Watson Machine Learning repository.

<a id="visualization"></a>
## 5. Predict locally and visualize

In this section you will learn how to score test data using loaded model and visualize the prediction results with the Plotly package.

### 5.1: Make a local prediction using previously loaded model and test data

Score the predict_data data set:

In [40]:
predictions = loadedModelArtifact.model_instance().transform(predict_data)

Preview the results by calling the *show()* method on the predictions DataFrame.

In [41]:
predictions.filter(predictions.prediction == 1.0).show()

+-------+---------+-----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+-----------+-----------+--------+----------+------------+--------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+--------------+--------------+-----+--------------------+--------------------+--------------------+----------+
|SEGMENT|PRVDR_NUM|CLM_PMT_AMT|HCPCS_CD_1|HCPCS_CD_2|HCPCS_CD_3|HCPCS_CD_4|HCPCS_CD_5|HCPCS_CD_6|HCPCS_CD_7|HCPCS_CD_8|HCPCS_CD_9|HCPCS_CD_10|HCPCS_CD_11|RESUBMIT|SEGMENT_IX|PRVDR_NUM_IX|CLM_PMT_AMT_IX|HCPCS_CD_1_IX|HCPCS_CD_2_IX|HCPCS_CD_3_IX|HCPCS_CD_4_IX|HCPCS_CD_5_IX|HCPCS_CD_6_IX|HCPCS_CD_7_IX|HCPCS_CD_8_IX|HCPCS_CD_9_IX|HCPCS_CD_10_IX|HCPCS_CD_11_IX|label|            features|       rawPrediction|         probability|prediction|
+-------+---------+-----------+----------+----------+----------+----------+----------+----------+----------+----------+---

Tabulate a count to see which claims will be resubmitted:

In [42]:
predictions.select("prediction").groupBy("prediction").count().show()

+----------+-----+
|prediction|count|
+----------+-----+
|       0.0| 1035|
|       1.0|  188|
+----------+-----+



### 5.2: Sample visualization of data with Plotly package

In this subsection you will explore prediction results with Plotly, which is an online analytics and data visualization tool.

**Example**: First, you need to install required packages. You can do it by running the following code. Run it only one time.<BR><BR>
!pip install plotly --user <BR>
!pip install cufflinks --user

Import Plotly and other required packages.

In [43]:
import sys
import pandas
import plotly.plotly as py
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.tools as tls
import plotly.graph_objs as go
init_notebook_mode(connected=True)
sys.path.append("".join([os.environ["HOME"]])) 

Convert the Apache Spark DataFrame to a Pandas DataFrame.

In [44]:
predictions_pdf = predictions.toPandas()

Plot a pie chart that shows the percentage of predicted resubmitted claims for the predict test dataset.

In [45]:
cumulative_stats = predictions_pdf.groupby(['prediction']).count()

fraud_data = [go.Pie(
            labels=cumulative_stats.index,
            values=cumulative_stats['label'],
    )]

fraud_layout = go.Layout(
    title='Resubmission claims distribution from test dataset',
)

fig = go.Figure(data=fraud_data, layout=fraud_layout)
iplot(fig)

With this data set, you might want to do some analysis of the probability of resubmission per claim code by using a bar chart.

In [46]:

resubmittedClaims = predictions_pdf.query('prediction == 1.0')
print(resubmittedClaims.SEGMENT + " " + 
      resubmittedClaims.PRVDR_NUM + " " + 
      resubmittedClaims.CLM_PMT_AMT + " " + 
      resubmittedClaims.HCPCS_CD_1 + " " +   
      resubmittedClaims.HCPCS_CD_2 + " " +   
      resubmittedClaims.HCPCS_CD_3 + " " +   
      resubmittedClaims.HCPCS_CD_4 + " " + 
      resubmittedClaims.HCPCS_CD_5 + " " +  
      resubmittedClaims.HCPCS_CD_6 + " " + 
      resubmittedClaims.HCPCS_CD_7 + " " +  
      resubmittedClaims.HCPCS_CD_8 + " " + 
      resubmittedClaims.HCPCS_CD_9 + " " + 
      resubmittedClaims.HCPCS_CD_10 + " " + 
      resubmittedClaims.HCPCS_CD_11
     
     )



3       1 0100NG 200 __NA__ __NA__ __NA__ __NA__ __NA_...
5       1 0100UP 100 __NA__ 96372 99282 __NA__ __NA__ ...
18      1 0300NP 40 __NA__ J2710 93005 __NA__ __NA__ _...
24      1 0300VH 3300 J2930 86140 84484 99211 __NA__ 8...
32      1 0400WD 600 __NA__ __NA__ __NA__ __NA__ 80048...
34      1 04018J 600 __NA__ 99214 66984 96367 __NA__ _...
37      1 0401PQ 200 __NA__ J3370 81003 J7030 __NA__ _...
45      1 0500DH 300 __NA__ __NA__ __NA__ __NA__ __NA_...
58      1 0502NA 0 __NA__ 85025 74170 __NA__ __NA__ __...
61      1 0502NA 50 __NA__ __NA__ __NA__ __NA__ __NA__...
77      1 0505QG 600 __NA__ __NA__ __NA__ __NA__ __NA_...
89      1 0513TN 20 __NA__ __NA__ __NA__ __NA__ __NA__...
90      1 0518NK 40 __NA__ 85610 99285 G0378 87186 757...
95      1 0518VN 70 __NA__ __NA__ __NA__ __NA__ __NA__...
103     1 0613SN 40 __NA__ __NA__ __NA__ __NA__ __NA__...
106     1 0700NU 0 __NA__ 45385 G0206 __NA__ 99283 J24...
122     1 1000AH 700 __NA__ 84450 __NA__ __NA__ __NA__...
132     1 1000

Based on bar plot you created, you might make the following conclusion: The following SEGMENT, PRVDR_NUM code has maximum chances of being a fraudulent claim

<a id="scoring"></a>
## 6. Deploy and score in a Cloud

In this section you will learn how to create online scoring and to score a new data record by using the Watson Machine Learning REST API. 
For more information about REST APIs, see the [Watson Machine Learning API Documentation](http://watson-ml-api.mybluemix.net/).

To work with the Watson Machine Leraning REST API, generate an access token:

In [47]:
import urllib3, requests, json

headers = urllib3.util.make_headers(basic_auth='{}:{}'.format(username, password))
url = '{}/v3/identity/token'.format(service_path)
response = requests.get(url, headers=headers)
mltoken = 'Bearer ' + json.loads(response.text).get('token')
print (mltoken)


Bearer eyJhbGciOiJSUzUxMiIsInR5cCI6IkpXVCJ9.eyJ0ZW5hbnRJZCI6IjExODM4NmJhLTQzMmQtNDE0Ny05NGFjLWQ4NTk3OGY0NDU1MCIsImluc3RhbmNlSWQiOiIxMTgzODZiYS00MzJkLTQxNDctOTRhYy1kODU5NzhmNDQ1NTAiLCJwbGFuSWQiOiIzZjZhY2Y0My1lZGU4LTQxM2EtYWM2OS1mOGFmM2JiMGNiZmUiLCJyZWdpb24iOiJ1cy1zb3V0aCIsInVzZXJJZCI6ImU3YWQzODY5LTkzNDYtNDZiYS1hNmFiLTI5M2I1NWI5OTAyOCIsImlzcyI6Imh0dHBzOi8vdXMtc291dGgubWwuY2xvdWQuaWJtLmNvbS92My9pZGVudGl0eSIsImlhdCI6MTU0Mzc3MDUyNSwiZXhwIjoxNTQzNzk5MzI1LCJjcmVhdGVkVGltZSI6MTU0Mzc3MDUyNX0.M1TLSHpzqe-Xi2GILrihbBOyEO-Ir-jGyYx-_XmTY2oXORVQI4b12ofGBW0kcA_dFKZ3RQo_tLH-seNini_Pogmu3_CFQjDxyjzUI6yQ-l9xTsdFSTIoffT6fEklUpnaktAFYFXKlledzIDRGvo7COf6u91O59tYwLUMvXV7VesghYn15J9yyFO-1lfHdqPbmLmcefpJPpxJ0XGRgoiw8itoyuuET4daXQCTTUJr7VlXDkdrj8Lyd_SIXEWN4XVOIsDmWdL4d3hGDdudfvfszi6qU6SVFcCwWDh7FJnR_QOdLn6qGcs6THf2PlVOKRpwR689UJCjysmsziiDR4RfDA


### 6.1: Create an online scoring endpoint

Now you can create an online scoring endpoint. Execute the following sample code that uses the **modelVersionHref** value to create the scoring endpoint to the Bluemix repository. Do only one deployment

response_online = requests.post(endpoint_online, json=payload_online, headers=header_online)
print response_online
print response_online.text

scoring_href = json.loads(response_online.text).get('entity').get('scoringHref')

In [48]:

endpoint_instance = service_path + "/v3/wml_instances/" + instance_id
header = {'Content-Type': 'application/json', 'Authorization': mltoken}

response_get_instance = requests.get(endpoint_instance, headers=header)
print (response_get_instance)
print (response_get_instance.text)

endpoint_published_models = json.loads(response_get_instance.text).get('entity').get('published_models').get('url')
print (endpoint_published_models)

header = {'Content-Type': 'application/json', 'Authorization': mltoken}
response_get = requests.get(endpoint_published_models, headers=header)

[endpoint_deployments] = [x.get('entity').get('deployments').get('url') 
for x in json.loads(response_get.text).get('resources') if x.get('metadata').get('guid') == saved_model.uid]

print(endpoint_deployments)

payload_online = {"name": "Claim Resubmission Prediction", "Description": "My Claim Resubmission Deployment", "type": "online"}
response_online = requests.post(endpoint_deployments, json=payload_online, headers=header)

scoring_url = json.loads(response_online.text).get('entity').get('scoring_url')
print (scoring_url)


<Response [200]>
{
  "metadata": {
    "guid": "118386ba-432d-4147-94ac-d85978f44550",
    "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550",
    "created_at": "2018-11-12T02:05:02.784Z",
    "modified_at": "2018-12-02T17:06:15.597Z"
  },
  "entity": {
    "source": "Bluemix",
    "published_models": {
      "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550/published_models"
    },
    "usage": {
      "expiration_date": "2019-01-01T00:00:00.000Z",
      "computation_time": {
        "current": 0
      },
      "model_count": {
        "limit": 1000,
        "current": 2
      },
      "prediction_count": {
        "current": 0
      },
      "gpu_count": {
        "limit": 8,
        "current": 0
      },
      "capacity_units": {
        "current": 0
      },
      "deployment_count": {
        "limit": 1000,
        "current": 9
      }
    },
    "tags": null,
    "plan_id": "0f2a3c2c-456b-

In [49]:
!pip install -U jsonpickle

tensorflow 1.3.0 requires tensorflow-tensorboard<0.2.0,>=0.1.0, which is not installed.
pyspark 2.3.0 requires py4j==0.10.6, which is not installed.


In [53]:
claim_scoring_endpoint='https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550/published_models/edb18ab1-f3fb-44c3-83fe-abc91ae6fc81/deployments/477b7b28-725d-4e44-a07f-4c4ba94ddc0a/online'
data = {}

data['fields']=["SEGMENT","PRVDR_NUM","CLM_PMT_AMT","HCPCS_CD_1",
                "HCPCS_CD_2","HCPCS_CD_3","HCPCS_CD_4","HCPCS_CD_5",
               "HCPCS_CD_6","HCPCS_CD_7","HCPCS_CD_8","HCPCS_CD_9","HCPCS_CD_10","HCPCS_CD_11"]
data['values'] = [["1", "01008J","50","77052","77057","__NA__","__NA__","__NA__","__NA__","__NA__","__NA__",   
                  "__NA__","__NA__","__NA__"]]
import jsonpickle
payload_online = jsonpickle.encode(data)
thawed = jsonpickle.decode(payload_online)
print (thawed)


response_online = requests.post(claim_scoring_endpoint, json=thawed, headers=header)
print (response_online)
print (response_online.text)


{'fields': ['SEGMENT', 'PRVDR_NUM', 'CLM_PMT_AMT', 'HCPCS_CD_1', 'HCPCS_CD_2', 'HCPCS_CD_3', 'HCPCS_CD_4', 'HCPCS_CD_5', 'HCPCS_CD_6', 'HCPCS_CD_7', 'HCPCS_CD_8', 'HCPCS_CD_9', 'HCPCS_CD_10', 'HCPCS_CD_11'], 'values': [['1', '01008J', '50', '77052', '77057', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__']]}
<Response [200]>
{
  "fields": ["SEGMENT", "PRVDR_NUM", "CLM_PMT_AMT", "HCPCS_CD_1", "HCPCS_CD_2", "HCPCS_CD_3", "HCPCS_CD_4", "HCPCS_CD_5", "HCPCS_CD_6", "HCPCS_CD_7", "HCPCS_CD_8", "HCPCS_CD_9", "HCPCS_CD_10", "HCPCS_CD_11", "SEGMENT_IX", "PRVDR_NUM_IX", "CLM_PMT_AMT_IX", "HCPCS_CD_1_IX", "HCPCS_CD_2_IX", "HCPCS_CD_3_IX", "HCPCS_CD_4_IX", "HCPCS_CD_5_IX", "HCPCS_CD_6_IX", "HCPCS_CD_7_IX", "HCPCS_CD_8_IX", "HCPCS_CD_9_IX", "HCPCS_CD_10_IX", "HCPCS_CD_11_IX", "features", "rawPrediction", "probability", "prediction"],
  "values": [["1", "01008J", "50", "77052", "77057", "__NA__", "__NA__", "__NA__", "__NA__", "__NA__", "__NA__", "__NA__", "__

In [55]:
claim_scoring_endpoint='https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550/published_models/edb18ab1-f3fb-44c3-83fe-abc91ae6fc81/deployments/477b7b28-725d-4e44-a07f-4c4ba94ddc0a/online'
data = {}

data['fields']=["SEGMENT","PRVDR_NUM","CLM_PMT_AMT","HCPCS_CD_1",
                "HCPCS_CD_2","HCPCS_CD_3","HCPCS_CD_4","HCPCS_CD_5",
               "HCPCS_CD_6","HCPCS_CD_7","HCPCS_CD_8","HCPCS_CD_9","HCPCS_CD_10","HCPCS_CD_11"]
data['values'] = [["1", "01008J","50","__NA__","77057","__NA__","__NA__","__NA__","__NA__","__NA__","__NA__",   
                  "__NA__","__NA__","__NA__"]]
import jsonpickle
payload_online = jsonpickle.encode(data)
thawed = jsonpickle.decode(payload_online)
print (thawed)


response_online = requests.post(claim_scoring_endpoint, json=thawed, headers=header)
print (response_online)
print (response_online.text)


{'fields': ['SEGMENT', 'PRVDR_NUM', 'CLM_PMT_AMT', 'HCPCS_CD_1', 'HCPCS_CD_2', 'HCPCS_CD_3', 'HCPCS_CD_4', 'HCPCS_CD_5', 'HCPCS_CD_6', 'HCPCS_CD_7', 'HCPCS_CD_8', 'HCPCS_CD_9', 'HCPCS_CD_10', 'HCPCS_CD_11'], 'values': [['1', '01008J', '50', '__NA__', '77057', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__', '__NA__']]}
<Response [200]>
{
  "fields": ["SEGMENT", "PRVDR_NUM", "CLM_PMT_AMT", "HCPCS_CD_1", "HCPCS_CD_2", "HCPCS_CD_3", "HCPCS_CD_4", "HCPCS_CD_5", "HCPCS_CD_6", "HCPCS_CD_7", "HCPCS_CD_8", "HCPCS_CD_9", "HCPCS_CD_10", "HCPCS_CD_11", "SEGMENT_IX", "PRVDR_NUM_IX", "CLM_PMT_AMT_IX", "HCPCS_CD_1_IX", "HCPCS_CD_2_IX", "HCPCS_CD_3_IX", "HCPCS_CD_4_IX", "HCPCS_CD_5_IX", "HCPCS_CD_6_IX", "HCPCS_CD_7_IX", "HCPCS_CD_8_IX", "HCPCS_CD_9_IX", "HCPCS_CD_10_IX", "HCPCS_CD_11_IX", "features", "rawPrediction", "probability", "prediction"],
  "values": [["1", "01008J", "50", "__NA__", "77057", "__NA__", "__NA__", "__NA__", "__NA__", "__NA__", "__NA__", "__NA__", "

<a id="summary"></a>
## 7. Summary and next steps     

 You successfully completed this notebook! You learned how to use Apache Spark machine learning as well as Watson Machine Learning for model creation and deployment. Check out our _[Online Documentation](https://apsportal.ibm.com/docs/content/analyze-data/ml-overview.html)_ for more samples, tutorials, documentation, how-tos, and blog posts. 

Copyright © 2017 IBM. This notebook and its source code are released under the terms of the MIT License.